### PR TITLE
Add a loading spinner + failure message for interop

### PR DESCRIPTION
**Pull request by lukebjerring**
## Description
Re-uses the `loadingCount` and `isLoading` properties in a `loading-state` behaviour component.

## Review Information
- Visit the auto-deployed env's interop tab. See a spinner.
- add `?label=foo`. See a failure message.

### CODE REVIEW

File: webapp/components/wpt-interop.html [hunk 1]
```diff
@@ -269,8 +289,11 @@
 
     async ready() {
       await super.ready();
+      this.loadingCount++;
+      const that = this;
```
Comment (mdittmer) [inline webapp/components/wpt-interop.html]: If you don't need the `arguments` object in `function(r) {...}` below, then prefer using `r => {...}` and dropping `that = this`.

File: webapp/components/wpt-interop.html [hunk 2]
```diff
@@ -308,6 +331,10 @@
       return metadata && metadata.test_runs;
     }
 
+    computeInteropNotFound(isLoading, interop) {
+      return !isLoading && !interop;
```
Comment (mdittmer) [inline webapp/components/wpt-interop.html]: nit: Don't rename `passRateMetadata` to `interop.

Won't this yield `interopNotFound = true` until `this.ready()` executes? Perhaps we should observe `isLoading` and check `this.passRateMetadata` exactly when `prevIsLoading = true && isLoading = false`.

File: webapp/components/wpt-interop.html [hunk 3]
```diff
@@ -7,7 +7,10 @@
 <link rel="import" href="../bower_components/polymer/polymer-element.html">
```
Comment (mdittmer) [inline webapp/components/wpt-interop.html]: Vain curiosity: Why bother with the rename?

File: webapp/components/wpt-interop.html (hunk 1 above)
Comment (lukebjerring) [reply webapp/components/wpt-interop.html]: Done.

File: webapp/components/wpt-interop.html (hunk 3 above)
Comment (mdittmer) [reply webapp/components/wpt-interop.html]: :+1:

File: webapp/components/wpt-interop.html (hunk 2 above)
Comment (lukebjerring) [reply webapp/components/wpt-interop.html]: Yeah, there was a small jank of 'not found' briefly. Done.

File: webapp/components/loading-state.html [hunk 4]
```diff
@@ -21,13 +21,28 @@
           isLoading: {
             type: Boolean,
             computed: 'computeIsLoading(loadingCount)',
+            observer: 'isLoadingChanged',
           },
+          /**
+           * wasLoading is true when something was actually loading, which
+           * distinguishes !isLoading for the not-yet-started case.
+           */
+          wasLoading: {
```
Comment (mdittmer) [inline webapp/components/loading-state.html]: I think this is more than we need (see comment below).

File: webapp/components/wpt-interop.html [hunk 5]
```diff
@@ -330,8 +333,8 @@
       return metadata && metadata.test_runs;
     }
 
-    computeInteropNotFound(isLoading, interop) {
-      return !isLoading && !interop;
+    computeInteropNotFound(wasLoading, isLoading, passRateMetadata) {
```
Comment (mdittmer) [inline webapp/components/wpt-interop.html]: Here and below. If we observe nothing but `isLoading`, then the observer function can see previous value and new value. Iff old value was `true` and new value is `false`, we can then manually check `this.passRateMetadata` this obviates the need for `wasLoading`. This also applies to `computeRunsNotFound` in `wpt-results.html`.

File: webapp/components/wpt-results.html [hunk 6]
```diff
@@ -324,7 +324,7 @@ <h3>
           },
           runsNotFound: {
             type: Boolean,
-            computed: 'computeRunsNotFound(isLoading, testRuns)',
+            computed: 'computeRunsNotFound(wasLoading, isLoading, testRuns)',
```
Comment (mdittmer) [inline webapp/components/wpt-results.html]: Ditto.

File: webapp/components/interop.html [hunk 7]
```diff
@@ -235,12 +251,16 @@
...
         onSearchCommit: Function,
+        interopLoadFailed: Boolean,
       };
     }
 
     constructor() {
       super();
       this.onSearchCommit = this.handleSearchCommit.bind(this);
+      this.onLoadingComplete = () => {
+        this.interopLoadFailed = !(this.passRateMetadata && this.allMetrics);
```
Comment (mdittmer) [inline webapp/components/interop.html]: It's not immediately obvious why `&& this.allMetrics` is included. Maybe add a comment about this predicate?

File: webapp/components/wpt-results.html [hunk 8]
```diff
@@ -321,17 +321,8 @@ <h3>
...
-          },
-          isLoading: {
-            type: Boolean,
-            computed: 'computeIsLoading(loadingCount)',
-          },
-          runsNotFound: {
-            type: Boolean,
-            computed: 'computeRunsNotFound(isLoading, testRuns)',
+          resultsLoadFailed: {
+            type:Boolean, value: false,
```
Comment (mdittmer) [inline webapp/components/wpt-results.html]: nits: Newline for each item; space after `type:`.

File: webapp/components/wpt-results.html [hunk 9]
```diff
@@ -412,17 +403,14 @@ <h3>
...
-      computeRunsNotFound(isLoading, testRuns) {
-        return !isLoading && !(testRuns && testRuns.length);
-      }
-
       constructor() {
         super();
         this.onSearchCommit = this.handleSearchCommit.bind(this);
+        this.onLoadingComplete = () => {
+          this.resultsLoadFailed =
+            !(this.testRuns && this.testRuns.length
```
Comment (mdittmer) [inline webapp/components/wpt-results.html]: Is this susceptible to transient effects? E.g., might `testRuns` start out as `undefined` or `[]` before being populated, and thus `(this && part)` is false, making `!(...)` true, and causing a transient error message to be shown?

File: webapp/components/wpt-results.html [hunk 10]
```diff
@@ -437,37 +425,45 @@ <h3>
...
       async fetchResults() {
         this.loadingCount++;
         try {
-          const response = await window.fetch(this.searchURL);
-          if (response.status !== 200) {
-            throw response;
-          }
-          const json = await response.json();
+          const json = await window.fetch(this.searchURL)
+            .then(async r => {
```
Comment (mdittmer) [inline webapp/components/wpt-results.html]: What change in behaviour are we trying to get by switching to `.then` inside an already async function? It seems like this change could just be a change in what gets thrown  when `response.status !== 200`.

File: webapp/components/interop.html (hunk 7 above)
Comment (lukebjerring) [reply webapp/components/interop.html]: The metadata is the runs and where to find the interop JSON blob, the allMetrics is the fetched JSON blob. If either fails, then the view is borked.

File: webapp/components/wpt-results.html (hunk 9 above)
Comment (lukebjerring) [reply webapp/components/wpt-results.html]: no; `testRuns` only gets set as the parsed JSON for the response, once.

File: webapp/components/wpt-results.html (hunk 10 above)
Comment (lukebjerring) [reply webapp/components/wpt-results.html]: Consistency with the exact same chunk of code in `interop`, which will make for a cleaner PR when the 2 get consolidated.

File: webapp/components/wpt-results.html (hunk 9 above)
Comment (mdittmer) [reply webapp/components/wpt-results.html]: Right. I thought this was a persistent binding, but I see now that it's inside `onLoadingComplete`, triggered when data is loaded. Ignore me.

File: webapp/components/wpt-results.html [hunk 11]
```diff
@@ -322,7 +322,8 @@ <h3>
             computed: 'computeAutocompleteURL(testRuns, search)',
           },
           resultsLoadFailed: {
-            type:Boolean, value: false,
+            type:Boolean,
```
Comment (mdittmer) [inline webapp/components/wpt-results.html]: nit: Still missing space after colon.

Comment (lukebjerring) [reply webapp/components/wpt-results.html]: sure; In future, just tweak the PR :)

**Pull request merged**